File: Pwr/Sem4/JezykiIParadygmatyProgramowania/Lista6/zad1/haskel/totient2.hs
import System.Environment (getArgs)

primeFactors :: Integer -> [Integer]
primeFactors n = factor n 2
  where
    factor 1 _ = []
    factor n f
        | f * f > n      = [n]
        | n `mod` f == 0 = f : factor (n `div` f) f
        | otherwise      = factor n (f + 1)

unique :: [Integer] -> [Integer]
unique [] = []
unique (x:xs) = x : unique (filter (/= x) xs)

totient2 :: Integer -> Integer
totient2 n =
    foldl (\acc p -> acc * (p - 1) `div` p) n (unique (primeFactors n))

main :: IO ()
main = do
    args <- getArgs
    print (totient2 (read (head args)))

File: Pwr/Sem4/JezykiIParadygmatyProgramowania/Lista6/zad1/ocaml/totient2.ml
let rec prime_factors n d =
  if n = 1 then []
  else if d * d > n then [n]
  else if n mod d = 0 then d :: prime_factors (n / d) d
  else prime_factors n (d + 1)

let rec unique = function
  | [] -> []
  | x :: xs -> x :: unique (List.filter (fun y -> y <> x) xs)

let totient2 n =
  let factors = unique (prime_factors n 2) in
  List.fold_left (fun acc p -> acc * (p - 1) / p) n factors

let () =
  let n = int_of_string Sys.argv.(1) in
  Printf.printf "%d\n" (totient2 n)